flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDbStateDataTransfer.java

```diff
@@ -0,0 +1,149 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.contrib.streaming.state;
+
+import org.apache.flink.core.fs.CloseableRegistry;
+import org.apache.flink.core.fs.FSDataInputStream;
+import org.apache.flink.core.fs.FSDataOutputStream;
+import org.apache.flink.core.fs.FileSystem;
+import org.apache.flink.core.fs.Path;
+import org.apache.flink.runtime.concurrent.FutureUtils;
+import org.apache.flink.runtime.state.IncrementalKeyedStateHandle;
+import org.apache.flink.runtime.state.StateHandleID;
+import org.apache.flink.runtime.state.StreamStateHandle;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+/**
+ * Data transfer utils for RocksDbKeyedStateBackend.
+ */
+public class RocksDbStateDataTransfer {
+
+	public static void transferAllStateDataToDirectory(
+		IncrementalKeyedStateHandle restoreStateHandle,
+		Path dest,
+		int resotreThreadNum,
+		CloseableRegistry closeableRegistry) throws Exception {
+
+		final Map<StateHandleID, StreamStateHandle> sstFiles =
+			restoreStateHandle.getSharedState();
+		final Map<StateHandleID, StreamStateHandle> miscFiles =
+			restoreStateHandle.getPrivateState();
+
+		transferAllDataFromStateHandles(sstFiles, dest, resotreThreadNum, closeableRegistry);
+		transferAllDataFromStateHandles(miscFiles, dest, resotreThreadNum, closeableRegistry);
+	}
+
+	/**
+	 * Copies all the files from the given stream state handles to the given path, renaming the files w.r.t. their
+	 * {@link StateHandleID}.
+	 */
+	private static void transferAllDataFromStateHandles(
+		Map<StateHandleID, StreamStateHandle> stateHandleMap,
+		Path restoreInstancePath,
+		int resotreThreadNum,
+		CloseableRegistry closeableRegistry
+		) throws Exception {
+
+		final ExecutorService executorService = Executors.newFixedThreadPool(resotreThreadNum);
+		List<CompletableFuture<Void>> futures = new LinkedList<>();
+		List<Closeable> closeables = new LinkedList<>();
+
+		try {
+			closeables.add(() -> executorService.shutdownNow());
+			closeableRegistry.registerCloseable(((LinkedList<Closeable>) closeables).getLast());
+
+			for (Map.Entry<StateHandleID, StreamStateHandle> entry : stateHandleMap.entrySet()) {
+				StateHandleID stateHandleID = entry.getKey();
+				StreamStateHandle remoteFileHandle = entry.getValue();
+
+				if (resotreThreadNum > 1) {
+					CompletableFuture<Void> future = CompletableFuture.runAsync(() -> {
+						try {
+							copyStateDataHandleData(new Path(restoreInstancePath, stateHandleID.toString()), remoteFileHandle, closeableRegistry);
+						} catch (IOException e) {
```

Comment:
the exception handling can go inside `copyStateDataHandleData`.
`new Path(restoreInstancePath, stateHandleID.toString())` can be a variable to reduce the line length.
I would create a list of `Runnable`'s:
```
Path path = new Path(restoreInstancePath, stateHandleID.toString());
runnables.add( -> copyStateDataHandleData(path, remoteFileHandle, closeableRegistry));
```
and then have separate code for `resotreThreadNum > 1` and `resotreThreadNum == 1`.
`executorService` currently is not used for `resotreThreadNum == 1` but created.